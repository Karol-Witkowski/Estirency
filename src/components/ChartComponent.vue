<template>
    <section>
      <header>
        <h3>
          {{ this.$store.state.baseCurrency.cc }}/{{ this.$store.state.targetCurrency.cc }}
          data from last year
        </h3>
      </header>
      <span v-if="$store.state.loaded">
        <datachart
          :chartdata="chartData"
          class="dataChart"
          :options="options"
        />
      </span>
    </section>
</template>

<script>
import axios from 'axios';
import datachart from '@/components/Chart.vue';

require('dotenv').config();

export default {
  name: 'ChartComponent',
  components: {
    datachart,
  },

  data() {
    return {
      actualDate: '',
      pastDate: '',
      historyData: [],
      chartData: {
        labels: [],
        datasets: [{
          borderColor: '#c7b6b6',
          pointBackgroundColor: '#a68b8b',
          pointHoverBorderWidth: 3,
          data: [],
        }],
      },
      options: {
        responsive: true,
        maintainAspectRatio: false,
        legend: {
          display: false,
        },
      },
    };
  },

  methods: {
    getDatePeriod() {
      const year = new Date().getFullYear();
      const month = new Date().getMonth() + 1;
      const day = new Date().getDate() - 1;

      if (month < 10 && day < 10) {
        this.actualDate = `${year}-0${month}-0${day}`;
        this.pastDate = `${year - 1}-0${month}-0${day}`;
      } else if (month > 9 && day < 10) {
        this.actualDate = `${year}-${month}-0${day}`;
        this.pastDate = `${year - 1}-${month}-0${day}`;
      } else if (month < 10 && day > 9) {
        this.actualDate = `${year}-0${month}-${day}`;
        this.pastDate = `${year - 1}-0${month}-${day}`;
      } else {
        this.actualDate = `${year}-${month}-${day}`;
        this.pastDate = `${year - 1}-${month}-${day}`;
      }
    },

    setData() {
      this.chartData.datasets[0].data = Object.keys(this.historyData)
        // eslint-disable-next-line max-len
        .map((k) => this.historyData[k][this.$store.state.targetCurrency.cc]).filter((e, i) => i % 20 === 0);
      this.chartData.labels = Object.keys(this.historyData).filter((e, i) => i % 20 === 0);
    },

    getData() {
      const historicalRate = 'https://api.exchangerate.host/timeseries?start_date=';

      if (this.$store.state.loaded === false) {
        axios.get(`${historicalRate}${this.pastDate}&end_date=${this.actualDate}&base=${this.$store.state.baseCurrency.cc}&symbols=${this.$store.state.targetCurrency.cc}`)
          .then((response) => {
            this.historyData = response.data.rates;
            this.setData();
            this.$store.state.loaded = true;
          })
          .catch((error) => error);
      }
    },
  },

  mounted() {
    this.$store.state.loaded = false;
    this.getDatePeriod();
    this.getData();
  },

  updated() {
    this.getData();
  },
};
</script>

<style src="@/stylesheets/ChartComponent.scss" lang="scss" scoped />
