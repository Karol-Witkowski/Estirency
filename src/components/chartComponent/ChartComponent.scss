.chart {
  float: left;
  display: flexbox;
  background-color: $backgrundColor;
  box-shadow: (2px 8px 30px -15px $primary);
  height: 580px;
  width: 735px;
  border: solid 1px $primary;
  border-radius: 10px;
  margin: 50px 30px 30px 30px;

  p {
    font-size: 1.3em;
    margin-top: 5%;
    text-transform: uppercase;
    overflow-wrap: nowrap;
    text-align: center;

    @media only screen and (max-width: 1520px) {
      margin-top: 4%;
    }

    @media only screen and (max-width: 650px) {
      font-size: 1.1em;
    }

    @media only screen and (max-width: 420px) {
      font-size: 1em;
    }

    @media only screen and (max-width: 380px) {
      font-size: 0.9em;
    }
  }

  #dataChart {
    height: 82.2%;
    margin: 3% 1% 3% 2%;

    @media only screen and (max-width: 1520px) {
      padding-bottom: -1%;
      margin-top: 0%;
    }

    @media only screen and (max-width: 1200px) {
      margin-top: 3.5%;
    }

    @media only screen and (max-width: 950px) {
      margin-top: 5%;
    }

    @media only screen and (max-width: 650px) {
      margin-top: 5%;
    }
  }
}