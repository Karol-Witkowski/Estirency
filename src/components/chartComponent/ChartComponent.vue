<template>
    <section>
      <header>
        <h3>
          {{ this.$store.state.baseCurrency.cc }}/{{ this.$store.state.targetCurrency.cc }}
          timeseries from last year
        </h3>
      </header>
      <span v-if="$store.state.loaded">
        <datachart
          :chartdata="chartData"
          class="dataChart"
          :options="options"
        />
      </span>
    </section>
</template>

<script>
import axios from 'axios';
import datachart from '@/components/chartComponent/chart/Chart.vue';

require('dotenv').config();

export default {
  name: 'ChartComponent',
  components: {
    datachart,
  },

  data() {
    return {
      actualDate: '',
      pastDate: '',
      historyData: [],
      chartData: {
        labels: [],
        datasets: [{
          borderColor: '#c7b6b6',
          pointBackgroundColor: '#a68b8b',
          pointHoverBorderWidth: 3,
          data: [],
        }],
      },
      options: {
        responsive: true,
        maintainAspectRatio: false,
        legend: {
          display: false,
        },
      },
    };
  },

  methods: {
    getDatePeriod() {
      const year = new Date().getFullYear();
      const month = new Date().getMonth() + 1;
      const day = new Date().getDate() - 1;

      if (month < 10 && day < 10) {
        this.actualDate = `${year}-0${month}-0${day}`;
        this.pastDate = `${year - 1}-0${month}-0${day}`;
      } else if (month > 9 && day < 10) {
        this.actualDate = `${year}-${month}-0${day}`;
        this.pastDate = `${year - 1}-${month}-0${day}`;
      } else if (month < 10 && day > 9) {
        this.actualDate = `${year}-0${month}-${day}`;
        this.pastDate = `${year - 1}-0${month}-${day}`;
      } else {
        this.actualDate = `${year}-${month}-${day}`;
        this.pastDate = `${year - 1}-${month}-${day}`;
      }
    },

    setData() {
      this.chartData.datasets[0].data = this.historyData
        .map((rate) => Number(rate.c));
      this.chartData.labels = this.historyData
        .map((date) => date.tm
          .slice(0, 10)
          .split('-')
          .reverse()
          .join('.'));
    },

    getData() {
      const historicalRate = 'https://fcsapi.com/api-v2/forex/history?symbol=';

      if (this.$store.state.loaded === false) {
        axios.get(`${historicalRate}${this.$store.state.baseCurrency.cc}/${this.$store.state.targetCurrency.cc}&period=1d&from=${this.pastDate}T12:00&to=${this.actualDate}T12:00&access_key=${process.env.VUE_APP_CURRENCY_API_KEY}`)

          .then((response) => {
            this.historyData = response.data.response
              .filter((e, i) => i % 20 === 0);
            this.setData();
            this.$store.state.loaded = true;
          })
          .catch((error) => error);
      }
    },
  },

  mounted() {
    this.$store.state.loaded = false;
    this.getDatePeriod();
    this.getData();
  },

  updated() {
    this.getData();
  },
};
</script>

<style src="./ChartComponent.scss" lang="scss" scoped />
