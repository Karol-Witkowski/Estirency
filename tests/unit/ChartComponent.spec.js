import { shallowMount } from '@vue/test-utils';
import ChartComponent from '@/components/ChartComponent.vue';
import axios from 'axios';

let wrapper;

jest.mock('axios', () => ({
  get: jest.fn(() => Promise.resolve({
    data: {
      response: {
        '2019-09-15': { cc: '0.2307' },
        '2019-09-17': { cc: '0.2432' },
      },
    },
  })),
}));

const currencies = [
  { cc: 'PLN', symbol: 'z\u0142', name: 'Polish zloty' },
  { cc: 'EUR', symbol: '\u20ac', name: 'European Euro' },
];

describe('ChartComponent.vue test', () => {
  wrapper = shallowMount(ChartComponent, {
    mocks: {
      $store: {
        state: {
          currency: currencies,
          baseCurrency: currencies[0],
          targetCurrency: currencies[1],
        },
      },
    },
  });

  it('Render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Check chart header', () => {
    expect(wrapper.find('h3[name=currencyHeader]').text()).toMatch('PLN/EUR');
  });

  it('Check that chart is not rendered', () => {
    expect(wrapper.find('[name=chart]').exists()).toBeFalsy();
  });

  it('Check axios call', () => {
    expect(wrapper.vm.historyData).toEqual({ '2019-09-15': '0.2307' }, { '2019-09-17': '0.2432' });
    expect(axios.get).toHaveBeenCalled();
    expect(axios.get).toHaveReturnedTimes(1);
  });

  it('Check that datachart exists when state.loaded = true', () => {
    wrapper = shallowMount(ChartComponent, {
      mocks: {
        $store: {
          state: {
            currency: currencies,
            baseCurrency: currencies[0],
            targetCurrency: currencies[1],
            loaded: true,
          },
        },
      },
    });

    expect(wrapper.find('[name=chart]').exists()).toBeTruthy();
  });

  it('Check setData method - map and set history data', () => {
    expect(wrapper.vm.chartData.labels).toEqual(['15.09.2019']);
    expect(wrapper.vm.chartData.datasets[0].data).toEqual([0.2307]);
  });
});
