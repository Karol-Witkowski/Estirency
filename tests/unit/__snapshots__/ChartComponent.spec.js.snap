// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartComponent.vue test Render correctly 1`] = `
<section>
  <header>
    <h3 name="currencyHeader">
      PLN/EUR
      data from last year
    </h3>
  </header>
  <!---->
</section>
`;
